README.md
<p align="center">
  <img src="https://raw.githubusercontent.com/frysztak/orpington-news/master/assets/logo.png" width="200px" height="200px" alt="logo">
  <h1 align="center">Orpington News</h2>
  <p align="center">
    Orpington News is a self-hosted RSS/Atom feed reader with PWA support.
    <br>
    <br>
    <a href="https://news-demo.orpington.software">🎮 View Demo</a>
  </p>
</p>

<p align="center">
  <a href="https://github.com/frysztak/orpington-news/releases" target="_blank">
    <img alt="GitHub release (latest SemVer)" src="https://img.shields.io/github/v/release/frysztak/orpington-news">
  </a>
  <a href="https://github.com/frysztak/orpington-news/pkgs/container/orpington-news" target="_blank">
    <img alt="Docker image size" src="https://ghcr-badge.deta.dev/frysztak/orpington-news/size">
  </a>
  <a href="https://github.com/frysztak/orpington-news/blob/master/LICENSE" target="_blank">
    <img alt="License" src="https://img.shields.io/github/license/frysztak/orpington-news">
  </a>
  <a href="https://codecov.io/gh/frysztak/orpington-news" target="_blank">
    <img alt="Codecov" src="https://codecov.io/gh/frysztak/orpington-news/branch/master/graph/badge.svg" />
  </a>
  <br />
  <a href="https://github.com/frysztak/orpington-news/actions/workflows/dev.yml" target="_blank">
    <img alt="Dev pipeline" src="https://github.com/frysztak/orpington-news/actions/workflows/dev.yml/badge.svg">
  </a>
  <a href="https://github.com/frysztak/orpington-news/actions/workflows/main.yml" target="_blank">
    <img alt="Main pipeline" src="https://github.com/frysztak/orpington-news/actions/workflows/main.yml/badge.svg?event=push">
  </a>
</p>

<p align="center">
  <a href="#features">Features</a> •
  <a href="#installation">Installation</a> •
  <a href="#development-quick-start">Development Quick Start</a>
  <br />
  <br />
  <a href="https://user-images.githubusercontent.com/217806/201395573-a9268808-4b53-45b5-ab22-0d7b68ec15e8.png" target="_blank">
    <img src="https://user-images.githubusercontent.com/217806/201395573-a9268808-4b53-45b5-ab22-0d7b68ec15e8.png" width="640px" alt="dark theme screenshot" />
  </a>
</p>

## Features

<ul>
  <li>RSS 1.0/RSS 2.0/Atom feed support<br/></li>
  <li>Nested collections drag-n-drop
    <video src="https://user-images.githubusercontent.com/217806/195674330-62585a3f-f430-42a0-aa77-5aa576752137.mp4"></video>
  </li>
  </li>
  <li>Light and dark theme<br/>
    <a href="https://user-images.githubusercontent.com/217806/201395742-d422dd26-c5e5-439e-a1eb-a6fcf0bd0bcc.png" target="_blank">
      <img src="https://user-images.githubusercontent.com/217806/201395742-d422dd26-c5e5-439e-a1eb-a6fcf0bd0bcc.png" width="390px" alt="light theme screenshot" />
    </a>
    <a href="https://user-images.githubusercontent.com/217806/201395573-a9268808-4b53-45b5-ab22-0d7b68ec15e8.png" target="_blank">
      <img src="https://user-images.githubusercontent.com/217806/201395573-a9268808-4b53-45b5-ab22-0d7b68ec15e8.png" width="390px" alt="dark theme screenshot" />
    </a>
  </li>
  <li>Mobile-friendly UI with PWA support
    <video src="https://user-images.githubusercontent.com/217806/195639397-26eac0d2-4527-472a-bf7e-01cbed5e0e42.mp4"></video>
  </li>
  <li>Syntax highlighting in code snippets</li>
  <li>Configurable fonts
    <br/><br/>
    Supported article fonts:
    <ul>
      <li>Nunito (default)</li>
      <li>Ubuntu</li>
      <li>Lato</li>
      <li>OpenDyslexic</li>
    </ul>
    <br/>
    Supported monospace fonts:
    <ul>
      <li>Source Code Pro (default)</li>
      <li>Ubuntu Mono</li>
      <li>Fira Mono</li>
      <li>OpenDyslexic Mono</li>
    </ul>
    <br/>
    Settings live preview:<br/>
    <a href="https://user-images.githubusercontent.com/217806/201395852-22023b63-738c-4779-8b4a-c8a71199aba3.png" target="_blank">
      <img src="https://user-images.githubusercontent.com/217806/201395852-22023b63-738c-4779-8b4a-c8a71199aba3.png" width="640px" alt="settings screenshot" />
    </a>
  </li>
  <li>Three collection layouts
    <br/>
    <ul>
      <li>
        Card<br/>
        <a href="https://user-images.githubusercontent.com/217806/201396193-b87cd218-5ec6-4885-afd7-a030db53da63.png" target="_blank">
          <img src="https://user-images.githubusercontent.com/217806/201396193-b87cd218-5ec6-4885-afd7-a030db53da63.png" width="440px" alt="card layout" />
        </a>
      </li>
      <li>Magazine<br/>
        <a href="https://user-images.githubusercontent.com/217806/201396481-fbb54761-a445-42c4-8e7e-e099fe13c4ef.png" target="_blank">
          <img src="https://user-images.githubusercontent.com/217806/201396481-fbb54761-a445-42c4-8e7e-e099fe13c4ef.png" width="440px" alt="card layout" />
        </a>
      </li>
      <li>List<br/>
      <a href="https://user-images.githubusercontent.com/217806/201396652-39fc0058-752d-460f-8cd7-38869edc8101.png" target="_blank">
          <img src="https://user-images.githubusercontent.com/217806/201396652-39fc0058-752d-460f-8cd7-38869edc8101.png" width="440px" alt="card layout" />
        </a>      
      </li>
    </ul>
  </li>
  <li>RSS feed auto-detection from page URL<br/>
    <video src="https://user-images.githubusercontent.com/217806/195674349-ab6a9a31-8dcd-4e2a-87f1-98cf2a16a0c2.mp4"></video>
  </li>
  </li>
  <li>OPML import</li>
  <li>Small Docker image size (40 MB uncompressed)</li>
</ul>

## Installation

You need two things: Docker and PostgresSQL instance.

Docker image (`ghcr.io/frysztak/orpington-news`) exposes the app on port `8000`. You also need to provide a handful of environmental variables:

- `APP_URL`
- `DB_HOST`
- `DB_PASS`
- `COOKIE_SECRET` - at least 64 characters long random string. If not set will use a fallback value. You can use `openssl rand -base64 64` to generate it.

For `DB_PASS` and `COOKIE_SECRET`, `_FILE` suffix is also supported. For example, `DB_PASS_FILE=./secrets/db_pass` will read database
password from file `secrets/db_pass`. All variables are described in [Wiki page](https://github.com/frysztak/orpington-news/wiki/Env-variables).

Sample Docker invocation:

```
docker run -it --name orpington-news --restart=always \
-e APP_URL=[ your public URL, including protocol ]  \
-e DB_HOST=[ your DB address ]  \
-e DB_PASS=[ your DB password ]  \
-p [ public port ]:8000  \
-d ghcr.io/frysztak/orpington-news
```

Sample `docker-compose.yml` is available [here](https://github.com/frysztak/orpington-news/blob/master/docker-compose.template.yml).

## Development quick start

Orpington News backend is written in Rust, frontend is written in TypeScript using Next.js. To run the app locally, you need:

- Node 18 (if you're using [nvm](https://github.com/nvm-sh/nvm), `.nvmrc` is provided)
- Rust toolchain (`rustup` is recommended)
- [Just](https://github.com/casey/just) command runner
- [bunyan-rs](https://github.com/LukeMathWalker/bunyan) log formatter
- [cargo watch](https://github.com/watchexec/cargo-watch)
- (optional) Docker, if you want to easily setup Postgres

Step-by-step guide:

1. Checkout the repo:

- `dev` branch, if you want latest and greatest
- `master` branch, if you want last tagged release

2. Start Postgres instance. You can do it any way you like, but for convenience, a `docker-compose.db.yml` file is provided. To run it:

```sh
$ docker compose -f backend/docker-compose.db.yml up -d
```

You should be able now to access pgAdmin at `http://localhost:8080`.

3. Install frontend dependencies:
```sh
$ cd frontend
$ npm ci
```

4. Orpington News uses `just` as a command runner, so running both frontend and API in watch mode is as simple as:
```sh
$ just
```

5. You should be able to access the app on `http://localhost:3000`, API will be available on `http://localhost:5000/api`. You'll need to create your user account. That's it!

## License

- [GNU GPL v3](http://www.gnu.org/licenses/gpl.html)
- Copyright 2021-2023


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.13.1] - 2023-08-16

### Fixed

-   Fixed Articles being unreliably marked as read

## [0.13.0] - 2023-08-11

### Added

-   Standalone Article page
-   Hotkeys
-   Previous/next Article buttons (mobile only)

### Fixed

-   Fix opening Article by URL
-   Fix previous/next Article calculations

### Removed

-   Swiping left/right on Article page

## [0.12.7] - 2023-07-14

### Added

-   Specific optimistic update for "mark all as read" operation for Collections with "unread" filter applied

### Fixed

-   Fix adding Collections without URL (#160)

### Changed

-   Update dependencies

## [0.12.6] - 2023-07-06

### Fixed

-   Fix User-Agent not always being set when fetching feeds (#149)

## [0.12.5] - 2023-07-03

### Fixed

-   Fix renaming Collections without URL (#145)
-   Fix URL validation (#149)

### Changed

-   Upgrade to Storybook 7

## [0.12.4] - 2023-06-25

### Added

-   ARM64 image

## [0.12.2] - 2023-03-27

### Fixed

-   Fix incorrect unread count being returned after refreshing collection
-   Fix session TTL not being refreshed
-   Fix API code coverage reporting incorrect numbers

### Added

-   Added `justfile`

### Changed

-   Changed from Server-Sent Events to WebSockets

## [0.12.1] - 2023-03-17

### Fixed

-   Fix HTML tags appearing in entry summary
-   Fix not being able to add Collection without URL
-   Fix parent collection items not being refetched when children were either marked as read or refreshed
-   Fix Article skeleton padding in expandable view

### Changed

-   Migrate to Next.js 13

## [0.12.0] - 2023-03-02

### Changed

-   Rewrite backend in Rust
    -   tiny Docker image, no Node.js runtime
    -   new task queue implementation
    -   more efficient feed refresh
    -   `COOKIE_SECRET` now must be at least 64 characters (if it's shorter, default value is used)
    -   dropped ARM64 image

## [0.11.1] - 2022-12-19

### Fixed

-   Fix OPML import
-   Fix missing Collection layout

## [0.11.0] - 2022-12-17

### Added

-   Performance: cache start page using service worker
-   Performance: use Tailwind for Drawer, Modal, Sidebar components
-   Respect "prefers reduced motion" setting
-   New syntax highlighter:
    -   detects language from HTML classes (rather than relying on Highlight.js' expensive language auto-detection)
    -   uses Prism.js and CSS-based styles

### Changed

-   Use Node.js 18

### Fixed

-   Fix disappearing Collections when reordering
-   Settings: fix mobile layout on some viewports

## [0.10.0] - 2022-12-04

### Added

-   Vertical panes layout
-   Expandable panes layout
-   Group by date: additional date buckets

### Changed

-   Drawer on mobile no longer takes 100% width
-   Made Article and Collection headers shorter on desktop

### Fixed

-   Icons for "Mark all as read"/"Mark as unread" are now different
-   List layout: improved scrolling performance on mobile

## [0.9.1] - 2022-11-28

### Fixed

-   Fixed crash when not using HTTPS ([#94](https://github.com/frysztak/orpington-news/issues/94))

## [0.9.0] - 2022-11-21

### Added

-   Collection item grouping (by date or by Collection)
-   Collection item sorting (newest/oldest first)

### Changed

-   Mark as read: relabel as "Mark all as read"

### Fixed

-   Fixed redundant fetching of empty collection pages
-   Mark as read: improve snappiness

## [0.8.0] - 2022-11-11

### Added

-   Collection filters (read, unread)
-   Collection item list: highlight active article
-   Collection item list: list layout
-   Article view: updated header
-   Add support for ETags, which significantly reduces collection refresh time

### Changed

-   Refactored how Home collection works internally
-   Changed collection page size to 30

### Fixed

-   Fixed missing loading state in Edit Collection modal

## [0.7.4] - 2022-10-20

### Fixed

-   Fixed changing layouts

## [0.7.3] - 2022-10-15

### Fixed

-   Fixed navigation when changing active collections on Firefox

## [0.7.2] - 2022-10-09

### Fixed

-   Improve UX of setting active collection and collapsing/expanding collections in high-latency scenarios
-   Fixed flickering when navigating to main page while not being logged-in

## [0.7.1] - 2022-10-06

### Added

-   Allow to mark Home collection as read
-   Improved UX of when marking article as read

### Changed

-   Hide article when changing active collection

### Fixed

-   Mark as read: properly refresh home collection

## [0.7.0] - 2022-09-25

### Added

-   Improved UX of when marking entire collection as read
-   Improved loading performance

### Changed

-   Migrated to React Query 4
-   Migrated to Slonik 31
-   Migrated to Jest 29
-   Migrated to TypeScript 4.8

## [0.6.4] - 2022-09-17

### Fixed

-   Fix handling of HTTP responses with `text/xml` content type
-   Fix feed detection for URLs with non-empty pathname (e.g. `https://example.com/blog`)

### Changed

-   Change default feed refresh interval to 6h
-   Collections list: increase overscan for smoother scrolling experience
-   Collection item list: show date an article was published

## [0.6.3] - 2022-08-26

### Fixed

-   Linux ARM64 build

## [0.6.2] - 2022-08-26

### Changed

-   Disable Linux ARM64 builds for now

## [0.6.1] - 2022-08-26

### Fixed

-   Linux ARM64 build

## [0.6.0] - 2022-08-26

### Added

-   Collections: implement OPML import
-   Collections: implement virtualization
-   Backend: better error logging
-   Backend: run feed updates in a separate thread
-   Backend: handle feeds in various encodings (that includes ISO-8859-2)

### Changed

-   Settings: reorganize some of the pages

## [0.5.1] - 2022-08-04

### Fixed

-   Fonts: use `latin-ext` variants
-   Card layout: fix little layout imperfection

### Changed


backend/src/config.rs
use secrecy::{ExposeSecret, Secret};
use sqlx::postgres::{PgConnectOptions, PgSslMode};
use std::env::var;
use std::fs::read_to_string;
use std::io;
use std::num::ParseIntError;
use std::str::ParseBoolError;
use tracing::warn;

#[derive(Debug)]
pub struct AppConfig {
    pub database: DBConfig,
    pub host: String,
    pub port: u16,
    pub cookie_secret: Secret<String>,
    pub app_url: String,
    pub disable_secure_cookie: bool,
}

#[derive(Debug)]
pub struct DBConfig {
    pub username: String,
    pub password: Secret<String>,
    pub port: u16,
    pub host: String,
    pub name: String,
}

#[derive(Debug)]
pub enum AppConfigError {
    ExclusiveOptions(String),
    PortParseError(ParseIntError),
    BoolParseError(ParseBoolError),
    IOError(io::Error),
}

impl From<ParseIntError> for AppConfigError {
    fn from(err: ParseIntError) -> Self {
        AppConfigError::PortParseError(err)
    }
}

impl From<ParseBoolError> for AppConfigError {
    fn from(err: ParseBoolError) -> Self {
        AppConfigError::BoolParseError(err)
    }
}

static DEFAULT_COOKIE_SECRET: &'static str =
    "gIFu/XtobXNB+iqbndgHualAe/a2k43w76VukhbG2s+q1CJst9BghwtQtD5/YUEwDmdGG8mea9QKcOjNU1ktSg==";

pub fn read_config() -> Result<AppConfig, AppConfigError> {
    let host = read_var("HOST", "0.0.0.0".to_string());
    let port = read_var("PORT", "5000".to_string()).parse()?;

    let cookie_secret =
        match read_var_from_file("COOKIE_SECRET", DEFAULT_COOKIE_SECRET.to_string())? {
            str if str.len() < 64 => {
                warn!("Cookie secret must be at least 64 characters, using default one");
                DEFAULT_COOKIE_SECRET.to_string()
            }
            str => str,
        };

    let app_url = var("APP_URL").expect("Missing APP_URL env var");
    let disable_secure_cookie = read_var("DISABLE_SECURE_COOKIE", "false".to_string()).parse()?;

    let db_pass = read_var_from_file("DB_PASS", "".to_string())?;
    let db_host = read_var_from_file("DB_HOST", "0.0.0.0".to_string())?;
    let db_port = read_var("DB_PORT", "5432".to_string()).parse()?;

    Ok(AppConfig {
        host,
        port,
        cookie_secret: Secret::new(cookie_secret),
        app_url,
        disable_secure_cookie,
        database: DBConfig {
            username: read_var("DB_USER", "postgres".to_string()),
            password: Secret::new(db_pass),
            host: db_host,
            port: db_port,
            name: read_var("DB_NAME", "postgres".to_string()),
        },
    })
}

fn read_var_from_file(var_name: &str, default_value: String) -> Result<String, AppConfigError> {
    let var_file_name = &format!("{}_FILE", var_name);
    let var_result = var(var_name);
    let var_file_result = var(var_file_name);

    match (var_result, var_file_result) {
        (Ok(_), Ok(__)) => Err(AppConfigError::ExclusiveOptions(format!(
            "Both {} and {} exist, please use just one",
            var_name, var_file_name
        ))),
        (Err(_), Ok(file_path)) => read_to_string(file_path)
            .map(|x| x.trim().to_string())
            .map_err(|err| AppConfigError::IOError(err)),
        (Ok(var_value), Err(_)) => Ok(var_value),
        (Err(_), Err(_)) => Ok(default_value),
    }
}

fn read_var(var_name: &str, default_value: String) -> String {
    let var_result = var(var_name);

    match var_result {
        Ok(var_value) => var_value,
        Err(_) => default_value,
    }
}

impl DBConfig {
    pub fn get_db_options(&self) -> PgConnectOptions {
        let ssl_mode = PgSslMode::Prefer;
        PgConnectOptions::new()
            .host(&self.host.to_string())
            .username(&self.username)
            .password(self.password.expose_secret())
            .port(self.port)
            .database(&self.name)
            .ssl_mode(ssl_mode)
    }
}
